### PR TITLE
Enable `requires_grad` on input embedding to train on top of frozen layers

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

## Motivation

In the context of `peft`, users currently needs to manually add a forward hook that enables gradient computation to the input after computing the embedding, for e.g. for `t5` one needs to call:
```python
def make_inputs_require_grad(module, input, output):
    output.requires_grad_(True)

model.get_input_embeddings().register_forward_hook(make_inputs_require_grad)
```

This PR makes the life easy for the users by wrapping this protocol in a single method `enable_input_require_grads` | Related: https://github.com/huggingface/peft/issues/80

cc @pacman100 

Wdyt @sgugger ? Maybe there is a better solution but not sure here, would love some guidance!

### CODE REVIEW

**Review by pacman100 — APPROVED** [review]
Hello, Thank you for adding these utils 🤗. To provide more context, when using PEFT method embedding layers have no trainable params, i.e., `lm_head` has no trainable params and in this situation gradient checkpointing will not backdrop saying that no trainable params found. To overcome this, outputs of the embedding layer are artificially made to require grads so that gradient checkpointing works as expected.

Reference issue: https://github.com/huggingface/peft/issues/80

The flow of hooks will be as follows, @sgugger can reconfirm:
1. module._hf_hook.pre_forward
2. old_forward()
3. module._hf_hook.post_forward
4. enable_input_require_grads() from the register_forward_hook

**Review by sgugger — APPROVED** [review]
This works for me, thanks!

**Comment by younesbelkada** [comment]
Thanks everyone! Merging!
